quaterion/loss/triplet_loss.py

```diff
@@ -124,3 +138,77 @@ def forward(
             triplet_loss = triplet_loss.mean()
 
         return triplet_loss
+
+    def _compute_xbm_loss(
+        self,
+        embeddings: Tensor,
+        groups: LongTensor,
+        memory_embeddings: Tensor,
+        memory_groups: LongTensor,
+    ) -> Tensor:
+        """Implement XBM loss computation for this loss.
+
+        Args:
+            embeddings: shape: (batch_size, vector_length) - Output embeddings from the
+                encoder.
+            groups: shape: (batch_size,) - Group ids associated with embeddings.
+            memory_embeddings: shape: (memory_buffer_size, vector_length) - Embeddings stored
+                in a ring buffer
+            memory_groups: shape: (memory_buffer_size,) - Groups ids associated with `memory_embeddings`
+
+        Returns:
+            Tensor: zero-size tensor, XBM loss value.
+        """
+        batch_size = embeddings.size(0)
+
+        # Compute similarity matrix
+        # shape: (batch_size, memory_buffer_size)
+        sim_mat = self.distance_metric.similarity_matrix(embeddings, memory_embeddings)
+
+        # split the positive and negative pairs
+        # shape: (batch_size, batch_size)
+        eyes_ = torch.eye(batch_size, dtype=torch.uint8).to(embeddings.device)
+
+        pos_mask = groups.expand(
+            memory_groups.shape[0], batch_size
+        ).t() == memory_groups.expand(batch_size, memory_groups.shape[0])
+        pos_mask = pos_mask.float()
+        neg_mask = 1 - pos_mask
+        pos_mask[:, :batch_size] = pos_mask[:, :batch_size] - eyes_
+
+        loss = list()
+        neg_count = list()
+
+        for i in range(batch_size):
+            pos_pair_idx = torch.nonzero(pos_mask[i, :]).view(-1)
+            if pos_pair_idx.shape[0] > 0:
+                pos_pair_ = sim_mat[i, pos_pair_idx]
+                pos_pair_ = torch.sort(pos_pair_)[0]
+
+                neg_pair_idx = torch.nonzero(neg_mask[i, :]).view(-1)
+                neg_pair_ = sim_mat[i, neg_pair_idx]
+                neg_pair_ = torch.sort(neg_pair_)[0]
+
+                select_pos_pair_idx = torch.nonzero(
+                    pos_pair_ < neg_pair_[-1] + self._margin
+                ).view(-1)
+                pos_pair = pos_pair_[select_pos_pair_idx]
+
+                select_neg_pair_idx = torch.nonzero(
+                    neg_pair_ > max(0.6, pos_pair_[-1]) - self._margin
```

Comment:
1) 0.6 shouldn't be a magic constant, consider assigning it to some class variable
2) do we actually want to choose `pos_pair_[-1]` - the best positive, or we want to choose the worst positive, i.e. `pos_pair_[0]` ?